docs/AsyncWrap/README.md

```diff
@@ -1,15 +1,202 @@
+but can also be inspected from the AsyncWrap hooks. When the handle is create
+using `new TCPConnectWrap()` the `init` hook is called.
+
+A `oncomplete` property is also set, this is the callback for when the connection is made or failed. Just before calling `oncomplete` the `before` hook
+is called, just after the `after` hook is called.
+
+The `TCP` handle works exactly the same way, except that the information
+is passed as arguments to a method `.connect` and the `onread` function
+is called multiply times, thus it behaves like an event. This also means that
+the `before` and `after` hooks are called multiply times.
+
+Thus one should expect the hooks be called in the following order:
+
+```javascript
+init // TCPConnectWrap
+init // TCP
+=== tick ===
+before // TCPConnectWrap
+after // TCPConnectWrap
+=== tick ===
+before // TCP
+after // TCP
+=== tick ===
+before // TCP
+after // TCP
+=== tick ===
+...
+```
+
+_tick_ indicates there is at least one tick between the text above and the text below.
+
+## The API
+
+At the moment there don't exist a high level API for AsyncWrap. It is simply exposed through `process.binding`:
+
+```javascript
+const asyncWrap = process.binding('async_wrap');
+```
+
+_Be warned that this API is not an official API and can change at any time, even if it's just patch
+update._
+
+To assign the hooks call:
+
+```javascript
+asyncWrap.setupHooks(init, before, after);
+function init(provider) { /* consumer code */ }
+function before() { /* consumer code */ }
+function after() { /* consumer code */ }
+```
+
+Note that calling `asyncWrap.setupHooks` again, will overwrite the previous
+hooks.
+
+#### Enable and disable
+
+Because there is a small performance penalty in just calling a noop function,
+AsyncWrap is not enabled by default. To enable AsyncWrap call:
+
+```javascript
+asyncWrap.enable();
+```
+
+Note that handle objects created before AsyncWrap is enabled will never
+be tracked, even after `.enable()` is called. Similarly when AsyncWrap is
+disabled, the handle objects there are already being tracked will continue
+to be tracked. Finally there are some cases where handle objects have a parent.
+For example a TCP server creates TCP sockets, thus the socket handles have the
+server handle as a parent. In this case AsyncWrap should be enabled before
+the server handle is created, for the TCP sockets to be tracked.
+
+Disabling AsyncWrap can be done with:
+
+```javascript
+asyncWrap.disable();
+```
+
+#### The hooks
+
+Currently there are 3 hooks: `init`, `before` and `after`. The function
+signatures are quite similar. In all cases the `this` variable is the handle
+object. Users may read properties from this object such as `port` and `address` in the `TCPConnectWrap` case, or set user specific properties.
+
+Finally the `init` has an extra argument called `provider`, this is an
+integer that refer to names defined in an `async.Providers` object map.
```

Comment:
Should make note that is isn't safe to iterate the `this` in the `init` callback. The object is partially constructed and there is at least one getter that'll segfault because the property isn't yet available. Though it is safe to assign values and read known properties.

Also note that with my previously mentioned patch the second argument in the init callback will be the parent object. Since it's impossible for all children to be created within the before/after callback scope, it can be used to perform better tracing.